www/js/lib/zimArchive.js

```diff
@@ -142,30 +142,71 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
+        var prefix = search.prefix;
         var that = this;
-        var prefixVariants = util.removeDuplicateStringsInSmallArray([prefix, util.ucFirstLetter(prefix), util.lcFirstLetter(prefix), util.ucEveryFirstLetter(prefix)]);
+        // Establish array of initial values that must be searched first
+        var startArray = [];
+        // Ensure a search is done on the string exactly as typed
+        startArray.push(prefix);
+        // Normalize any spacing and make string all lowercase
+        prefix = prefix.replace(/\s+/g, ' ').toLocaleLowerCase();
+        // Add lowercase string with initial uppercase (this is a very common pattern)
+        startArray.push(prefix.replace(/^./, function (m) {
+            return m.toLocaleUpperCase();
+        }));
+        // Get the full array of combinations to check number of combinations
+        var fullCombos = util.removeDuplicateStringsInSmallArray(util.allCaseFirstLetters(prefix, 'full'));
+        // Put cap on exponential number of combinations (five words = 3^5 = 243 combinations)
+        search.type = fullCombos.length < 200 ? 'full' : 'basic';
+        // We have to remove duplicate string combinations because util.allCaseFirstLetters() can return some combinations
+        // where uppercase and lowercase combinations are exactly the same, e.g. where prefix begins with punctuation
+        // or currency signs, for languages without case, or where user-entered case duplicates calculated case
+        var prefixVariants = util.removeDuplicateStringsInSmallArray(
+            startArray.concat(
+                // Get basic combinations first for speed of returning results
+                util.allCaseFirstLetters(prefix).concat(
+                    search.type === 'full' ? fullCombos : []
+                )
+            )
+        );
         var dirEntries = [];
+        var inProgressResults = [];
+
         function searchNextVariant() {
+            // If user has initiated a new search, cancel this one
+            if (search.state === 'cancelled') return;
             if (prefixVariants.length === 0 || dirEntries.length >= resultSize) {
-                callback(dirEntries);
+                search.state = 'complete';
+                callback(dirEntries, search);
                 return;
             }
+            // Dynamically populate list of articles
+            search.state = 'interim';
+            if (!noInterim) callback(dirEntries, search);
             var prefix = prefixVariants[0];
             prefixVariants = prefixVariants.slice(1);
-            that.findDirEntriesWithPrefixCaseSensitive(prefix, resultSize - dirEntries.length, function (newDirEntries) {
-                dirEntries.push.apply(dirEntries, newDirEntries);
-                searchNextVariant();
+            that.findDirEntriesWithPrefixCaseSensitive(prefix, resultSize - dirEntries.length, search, function (newDirEntries, interim) {
+                if (search.state === 'cancelled') return;
```

Comment:
It would be better to `callback ([], search)` here, i.e. with an empty array, and let the callback in app.js handle the cancellation.